src/components/ContactsAdd.js

```diff
@@ -1,39 +1,101 @@
-import { useState } from "react"
+import { useState } from "react";
 import { useNavigate } from "react-router-dom";
 
 function ContactsAdd(props) {
+  const initialNewContact = {
+    firstName: "",
+    lastName: "",
+    street: "",
+    city: "",
+  };
 
-  // setContacts and contacts must be passed as props
-  // to this component so new contacts can be added to the
-  // state
-  const { setContacts, contacts } = props
+  const { setContacts, contacts } = props;
+  const [newContact, setNewContact] = useState(initialNewContact);
 
-  //TODO: Implement controlled form
-  //send POST to json server on form submit
+  function handleChange(event) {
+    const { name, type, value } = event.target;
+    if (name === "firstName" && type === "text") {
+      setNewContact({ ...newContact, firstName: value });
+    }
+    if (name === "lastName" && type === "text") {
+      setNewContact({ ...newContact, lastName: value });
+    }
+    if (name === "street" && type === "text") {
+      setNewContact({ ...newContact, street: value });
+    }
+    if (name === "city" && type === "text") {
+      setNewContact({ ...newContact, city: value });
+    }
+  }
+
+  function handleSubmit(event) {
+    event.preventDefault();
+    fetch("http://localhost:4000/contacts", {
+      method: "POST",
+      headers: {
+        "Content-Type": "application/json",
+      },
+      body: JSON.stringify(newContact),
+    })
+      .then((res) => res.json())
+      .then((result) => {
+        setContacts([...contacts, result]);
+        event.target.reset();
```

Comment:
what does this do? I've seen it in a couple people's code but did not include it myself